KEP-1714: Fair Sharing (#1773)
kubernetes-sigs/kueue: keps/1714-fair-sharing/README.md

  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
    - [Story 3](#story-3)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Share value function and weights](#share-value-function-and-weights)
  - [Preemption algorithm](#preemption-algorithm)
    - [Choosing ClusterQueues to preempt from](#choosing-clusterqueues-to-preempt-from)
    - [Choosing workloads from ClusterQueues for preemption](#choosing-workloads-from-clusterqueues-for-preemption)
    - [Algorithmic Optimization](#algorithmic-optimization)
    - [Minimum number of preemptions](#minimum-number-of-preemptions)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha](#alpha)
    - [Beta](#beta)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary
This KEP introduces weight-based fair sharing of unused resources across 
needing ClusterQueues, respecting borrowing and lending limits, multi-level
hierarchy and preferences of cohorts to distribute unused resources 
Internally.

## Motivation

> Currently unused resources are allocated on the "first come first served" basis.

Author: Added.

to borrowing, however in organizations with no strict budget enforcing, priorities
could be overused (if priorities could guarantee more resources, then everyone would
use the highest one, breaking the whole idea altogether). Thus a more comprehensive
solution of distributing unused resources is needed.

### Goals
* Create a mechanism to enforce fair sharing of resources. Two equally important
sub-organizations (Cohorts or ClusterQueues), placed in the similar spots in the
whole organization (hierarchy of Cohorts), actively competing for the same
resources (having workloads needing more than nominal quota), should 
be given a similar amount of resources. And any pre existing imbalances should
be timely resolved by preemption.

* Allow system admin to indicate that a sub-organization is more important than
their siblings, and should get proportionally more resources than others located
under the same parent. 

* When distributing resources, prioritize needs of CQ within the same
sub-organization and only after they have been fulfilled, proceed with distribution
outside of the suborganization.

* When enforcing fair sharing, ignore workload priorities unless:

> ** The workload's priority is above admin-defined high priority. Super high

Author: IMHO, only slightly. There should be a way to bypass fair-sharing without turning it off completely.

unused capacity) should be resolved in favor of fair sharing than maximizing the
utilization (the big greedy workload should be preempted to admit smaller
workloads from other CQ that consume only their fair share).

### Non-Goals
* Use historical data (for example CQ A used a lot of shared capacity for the last
week, so now it should get less because others, who didn't need anything then, have
pending workloads). Fair sharing should be based on point-in-time situation, although,
ideally it should be expandable to support history-based fair sharing without major redesign.

* Enable fair sharing only for some part of the resources or Cohort hierarchy.
Fair sharing will be a global switch (at least initially).

* Maximize utilization at the cost of fair sharing.

## Proposal
Introduce a global fair sharing mechanism that is based on preemptions. As long
as there are some free and accessible resources in the cohort hierarchy, Kueue
will admit workloads without any limits. However, once the capacity is gone,
new workloads from Cohorts/CQ that have not received their fair share yet will
be permitted to preempt workloads from Chorots/CQ that received more than their

> fair share would otherwise be (or more than the Cohort/CQ that is in need).

Reviewer: Can we define at this point what "fair share" is? In particular, is it proportional to "nominalQuota", what is the unit of the values?

I assume it is defined later, but it seems reasonable to define it early, not necessarily fully formally, so that the reader has some intuition what this is already. If not feasible to define here maybe at least reference with the line to the place where the concept is defined later.

Author: Added.

Additionally, we will introduce a reclaim mechanism at the Cohort level. Cohorts
that are in need, but whose consumption didn’t reach their total aggregated

> nominal quota will be able to preempt workloads from Cohorts/CQ that are over

Reviewer: What about ClusterQueues which reached the total aggregated quota, aren't they able to preempt? This scenario was one of the motivations for https://github.com/kubernetes-sigs/kueue/tree/main/keps/1337-preempt-within-cohort-while-borrowing, in which nominalQuota for ClusterQueues is 0 (so already reached), yet they are able to preempt. I'm wondering if we are not losing this scenario here.

within its own nominal quota (very similar to what we currently have for CQ
within the single level Cohort structure).

We will add an optional weight field to both Cohorts and CQ. The weight will
indicate how to fair share resources between sub-organizations (CQs or Cohorts)
under the same Cohort.

Fair sharing will be configured for the whole cluster, using the configuration
file. In Alpha it will be just a feature gate.

### User Stories (Optional)

#### Story 1
I have 2 teams in my department in some US-based company, one on the east coast
(TeamE), one on the west coast (TeamW). Both come to work at 9 am their local time.
And immediately after that they start submitting their research jobs for execution
in their assigned CQs. Each of them have a bit of guaranteed resources, however they
both rely on a company-wide pool or resources that should be fairly shared. 

TeamE can submit as many workloads and consume as many resources as they can while
TeamW is not a work and doesn’t need resources. However, once they arrive, some of
the already submitted workloads from TeamE may be preempted(preferably the least

> important) to ensure equal extra space (irregardless of their given quota) for both teams. 

Reviewer: ```suggestion
important) to ensure equal extra space (regardless of their given quota) for both teams. 
```
nit, unless you like to use fancy words :)

submits their workloads to the appropriate CQ: 1A, 1B, 1C, 2A, 2B, 2C. Each of the
teams/CQs is given some guaranteed resources within the big ML cluster. Each team
is also allowed to borrow as many resources from other teams as they want, as long
as they are not needed by the owner. However, due to internal accounting, company
internal politics, etc, free resources should first be distributed within the
department. For example:

* if CQ 1A has free resources, and both 1B and 1C are in need of extra resources 
to run their workloads, then 2A, 2B and 2C should not be given a share of 1A,
that would block any workload from 1B or 1C from running. 

* If CQ 1A has free resources and 1B and 1C do not need any of those to start
a workload (there are no pending workloads or they are too big anyway) then 2A, 2B,
2C should be given an equal share of the resources that correspond to their needs.

#### Story 3
I have 3 cohorts inside my organization (top cohort): CS, C1 and C2. CS contains
a single CQ that is given 300 GPU nominal quota. C1 contains 3 CQ: 1A, 1B and 1C,
among which 1C is considered to be 3 times more important than others. C2 contains

> only one CQ: 2A.  Each of the queues 1A, 1B, 1C and 2C has a long backlog of jobs

Reviewer: ```suggestion
only one CQ: 2A.  Each of the queues 1A, 1B, 1C and 2C has a long backlog of jobs
```
```suggestion
only one CQ: 2A.  Each of the queues 1A, 1B, 1C and 2A has a long backlog of jobs
```
?

to execute. I want to distribute the resources from CS in the following way:

* Both C1 and C2 get 150 GPUs each, because they are considered equally important.
* Inside C1, out of these 150 GPUs, 1A gets 30, 1B gets 30 and 1C gets 90 GPUs. 
1A and 1B are equally important but 1C is 3x more important than 1A and 1B.
* Inside C2 there is just a single CQ, so it gets all 150 GPUS.

### Risks and Mitigations

* Fair sharing may increase the number of preemptions in the system vs current
state where the first workload to acquire unused resources keeps them until it
finishes. Mitigations include:
    * Introduce minimum execution time before workloads can be preempted for fair sharing.

>     * Introduce delayed fair share enforcement - new workloads have to wait a bit before preempting others to get their share.

Reviewer 2: Can you add the proposed configuration API?

Author: This can be updated once we will be implementing the sub-feature.

Reviewer 3: If we skip these options in the first iteration, I'm ok with this.
We can update this KEP before the second iteration.

* Fair sharing may decrease utilization of the unused resources while attempting to
distribute them fairly, vs provide the tighties bin-packing. To avoid these scenarios,
users should prefer to run massive workloads under nominal quotas.

> ## Design Details

Reviewer: Is the design based on some state-of-the-art approaches? If so, even if modified (or taken as inspiration), it could be helpful to provide some references to justify the design choice, so that we are not re-inventing the wheel.

Author: The approach is not (up to my best knowledge) directly based on any existing solution. There are some similarities, but adding references may not necessarily benefit the reader.

The enforcement of fair-share revolves around the idea of a share value per CQ or
Cohort and mechanisms to keep the share values among siblings in the hierarchy to
be as equal as possible. This is achieved via two mechanisms:

> * During admission, prioritize admitting workload from the ClusterQueues and cohorts that have the lowest share value.

Reviewer 4: One general question is: since DRF will try to distribute the workloads to different resource pool(CQ) as a result of favoring the smaller share value, but preemption will try to evict the borrowers, isn't this conflict, which means we'll evict more than we're today.

Reviewer 2: There is no conflict. See the rules S2-a and S2-b in `Choosing workloads from ClusterQueues for preemption`.

It's easy to see that S2-a increases the value of the smallest CQ but it remains being the one with smallest share.

S2-b has a typo, but see my comment below.

That said, fair sharing will indeed increase the number of preemptions overall, but the algorithm should reach a steady state. In other words, there wouldn't be flapping.

* Preempt workloads from ClusterQueues and cohorts that have the highest share value.

### Share value function and weights

> The value function is a variation of DRF:

Reviewer: Would be good to expand on the DRF acronym, if this is some state-of-the-art approach we could mention some reference we draw inspiration from (as mentioned above).

Reviewer 2: These 2 papers:
- https://amplab.cs.berkeley.edu/wp-content/uploads/2011/06/Dominant-Resource-Fairness-Fair-Allocation-of-Multiple-Resource-Types.pdf
- https://dash.harvard.edu/bitstream/handle/1/11956916/Parkes_BeyondDominant.pdf;jsessionid=AC0D06C2CC07C693BD42008D7AE25D99?sequence=1
I agree, worth adding.

Author: Added.

For a given resource r provided by a ClusterQueue or cohort c, we calculate T_r as the total
requests consumed by the Workloads for that resource in that CQ or cohort, independent of the
flavor, that are above the nominal quota. The value for a resource is the ratio of T_r and the

> total nominal quotas in the hierarchy of the parent of C.

Reviewer 2: ```suggestion
total nominal quotas (or lendingLimits, if defined) in the hierarchy of the parent of C.
```

> The value for the CQ or cohort is the maximum among the values for each resource, divided by the weight, if defined.

Reviewer: "The **value** for the CQ or cohort" - is this the fair share value? This paragraph feels quite abstract, I think it would be helpful to back it up with a small example so that we can have some intuition.

Weights will be added to ClusterQueueSpec and CohortSpec in the following optional struct:

```go
type struct FairSharing {

>      Weight 	*Quantity  // Default is 1. 

Reviewer 2: line 200

}
```

### Preemption algorithm

The algorithm could be summarized as follows:

```
AdmitWorkload(x CQ, w Workload)

>  While w does not fit:

Reviewer: Is this algorithm guaranteed to stop? In particular is it possible there are no workloads satisfying this "Find workloads z that could be preempted while satisfying fairness", while the workloads still does not fit?

Reviewer 2: ```suggestion
 While w does not fit and there are workloads that can still be preempted:
```

* The mechanism should be compatible with all existing or proposed Kueue features,
in particular:

** Borrowing limits 
** Lending limits
** Guaranteed/nominal quota
** Hierarchical cohorts

* Fair sharing should not limit Kueue scalability. Kueue, with fair sharing enabled,
should be able to handle >1k ClusterQueues, >100 Cohorts and >10k workloads (that
are either running or queued) within a single hierarchical organization.

* The proposed system should be hard to game, for example by creating big workloads
that consume all capacity.

* Fair sharing enforcement should not significantly decrease overall utilization,
however, pathological situations (like a single workload consuming all otherwise
unused capacity) should be resolved in favor of fair sharing than maximizing the
utilization (the big greedy workload should be preempted to admit smaller
workloads from other CQ that consume only their fair share).

### Non-Goals
* Use historical data (for example CQ A used a lot of shared capacity for the last
week, so now it should get less because others, who didn't need anything then, have
pending workloads). Fair sharing should be based on point-in-time situation, although,

> ideally it should be expandable to support history-based fair sharing without major redesign.

Author: Added.

1A and 1B are equally important but 1C is 3x more important than 1A and 1B.
* Inside C2 there is just a single CQ, so it gets all 150 GPUS.

### Risks and Mitigations

* Fair sharing may increase the number of preemptions in the system vs current
state where the first workload to acquire unused resources keeps them until it
finishes. Mitigations include:
    * Introduce minimum execution time before workloads can be preempted for fair sharing.
    * Introduce delayed fair share enforcement - new workloads have to wait a bit before preempting others to get their share.

* Fair sharing may decrease utilization of the unused resources while attempting to
distribute them fairly, vs provide the tighties bin-packing. To avoid these scenarios,
users should prefer to run massive workloads under nominal quotas.

## Design Details

The enforcement of fair-share revolves around the idea of a share value per CQ or
Cohort and mechanisms to keep the share values among siblings in the hierarchy to
be as equal as possible. This is achieved via two mechanisms:

* During admission, prioritize admitting workload from the ClusterQueues and cohorts that have the lowest share value.

> * Preempt workloads from ClusterQueues and cohorts that have the highest share value.

Reviewer 2: Let's add:

The algorithms take a decision from root-to-leaves. This favors admission from sub-organizations that, as a whole, still have unused resources, or favors preemptions from sub-organizations that, as a whole, are borrowing too much.

### Share value function and weights

The value function is a variation of DRF (see
[1](https://amplab.cs.berkeley.edu/wp-content/uploads/2011/06/Dominant-Resource-Fairness-Fair-Allocation-of-Multiple-Resource-Types.pdf),
[2](https://dash.harvard.edu/bitstream/handle/1/11956916/Parkes_BeyondDominant.pdf;jsessionid=AC0D06C2CC07C693BD42008D7AE25D99?sequence=1)):

For a given resource r provided by a ClusterQueue or cohort c, we calculate T_r as the total
requests consumed by the Workloads for that resource in that CQ or cohort, independent of the
flavor, that are above the nominal quota. The value for a resource is the ratio of T_r and the

> total nominal quotas (or lendingLimits, if defined) in the hierarchy of the parent of C.

Reviewer 2: Let's add another paragraph:

Note that the share value for a sub-organization (a node in the tree) is independent of the share value for its children. In other words, the calculation of the share value only depends on the sum of usages and sum of nominal quotas in the sub-organization.

Author: Done.

The value for the CQ or cohort is the maximum among the values for each resource, divided by the weight, if defined.

> Weights will be added to ClusterQueueSpec and CohortSpec in the following optional struct:

Reviewer 3: When we set FairSharing configuration, what happens?

Reviewer 3: This is my bad.
I wanted to say "When we set FairSharing configuration to ClusterQueue without Cohort, what happens?"

Reviewer 2: Nothing. Fair sharing only applies above CQs.

Reviewer 3: You meant that we will add any validation?

> ```go

Reviewer 3: As I can remember, we have a plan to have DRF as another queueingStarategy like this: https://docs.google.com/document/d/1VQ0qxWA-jwgvLq_WYG46OkXWW00O6q7b1BsR_Uv-acs/edit?usp=sharing

So, did you compare the pros and cons of the following options?

1. Exptend CohortSpec and ClusterQueueSpec (current your approach)
2. Introduce a new queueing strategy, "DRF", and have a new CRD, "FairSharing".

@mwielgus @alculquicondor

Reviewer 3: That makes sense.
Recording this discussion as an Alternative approach might be worth it, right?

When trying to preempt some workloads to admit workload w to CQ x we consider all other
CQs that are in the same hierarchy. For each of such CQ y we check whether scenario S1
or S2 occur. To do that we compare sub-organizations that are rooted in AlmostLCA(x,y) and AlmostLCA(y,x).

![LCA diagram](lca.png)

First of all we validate if y is not using its local nominal quota, somewhere on the path.
If y was using some local nominal quota that lies entirely under AlmostLCA(y,x) then x should
not be allowed to preempt any of its workload. So we check whether both y and all cohorts on
the way up to AlmostLCA(y,x) are borrowing. If that happens then y may have something to
do with x being unable to admit the workload. 

We trigger scenario S1 if AlmostLCA(x,y) is not borrowing but x still has too little
resources. We put all y’s workloads on a list of potential preemptions, out of which
we will later (once we check all other CQs) select one that is most reasonable. 

We trigger scenario S2 if AlmostLCA(x,y) is borrowing. We check whether AlmostLCA(x,y)
and AlmostLCA(y,x) are borrowing roughly the same amount. To do this we use a value 

> function (to be defined later) that returns a value of the borrowed capacity.

Reviewer 2: oops, the value is already defined above

Author: Fixed.

#### Choosing workloads from ClusterQueues for preemption

For each workload z in y we check whether if:

[S2-a] value of AlmostLCA(y,x) without z is still higher than value of AlmostLCA(x,y)
with admitted workload x. Y’s sub-orgs will still be better than X’s sub-org after we
preempt z and admit w, thus z is a reasonable candidate to re-balance fair sharing.

> [S2-b] value of AlmostLCA(y,x) is lower than AlmostLCA(x,y) with admitted workload x.

Reviewer 2: this seems to be a typo
```suggestion
[S2-b] value of AlmostLCA(y,x) is strictly higher than AlmostLCA(x,y) with admitted workload x.
```

Also clarify that this implies that, using this rule, the maximum share value in the cluster monotonically decreases, eventually stopping.
